src/poseidon_gencontract.js

```diff
@@ -117,29 +116,29 @@ export function createCode(nInputs) {
 
     saveM();
 
-    C.push("0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001");  // q
+    C.push("0x0800000000000011000000000000000000000000000000000000000000000001");  // q this is the big prime in starknet
 
     // Load t values from the call data.
     // The function has a single array param param
     // [Selector (4)] [item1 (32)] [item2 (32)] ....
     // Stack positions 0-nInputs.
-    for (let i=0; i<nInputs; i++) {
-        C.push(0x04+(0x20*(nInputs-i-1)));
+    //    C.push(0); // per starkwares implementation ->    https://github.com/starkware-libs/cairo-lang/blob/12ca9e91bbdc8a423c63280949c7e34382792067/src/starkware/cairo/common/poseidon_hash.py#L31
+    for (let i = 0; i < nInputs; i++) {
+        C.push(0x04 + (0x20 * (nInputs - i - 1)));
         C.calldataload();
     }
 
-    C.push(0);
 
-    for (let i=0; i<nRoundsF+nRoundsP; i++) {
+    for (let i = 0; i < nRoundsF + nRoundsP; i++) {
         ark(i);
-        if ((i<nRoundsF/2) || (i>=nRoundsP+nRoundsF/2)) {
-            for (let j=0; j<t; j++) {
+        if ((i < nRoundsF / 2) || (i >= nRoundsP + nRoundsF / 2)) {
+            for (let j = 0; j < t; j++) {
                 sigma(j);
             }
         } else {
-            sigma(0);
+            sigma(2);
```

Comment:
a bit tricky to follow :) why changing to `sigma(2)` here why it the code above didn't change?

Reply:
ok I think it's related to changes below with `C.push("0x40")` and `C.push("0x60")`, right?

Reply:
Actually no, sigma(0) -> sigma(2) change corresponds to the implementation detail here https://github.com/starkware-libs/cairo-lang/blob/12ca9e91bbdc8a423c63280949c7e34382792067/src/starkware/cairo/common/poseidon_utils.py#L85
 or may the GO implementation is clearer -> https://github.com/NethermindEth/juno/blob/8aad125ce11ce4bfc1df71d253a45f71a259ab16/core/crypto/poseidon_hash.go#L22. I will add comments, thanks so much for pointing out. This is dense and needs a bit of clarity. These links will help. The last C.push("0x40") C.push("0x40")  is part of the change where I have changed this from Poseidon(X,Y)=> A to Poseidon3(X,Y,Z) => A,B,C. Essentially a transformation from 3 element input to 3 element output.

Reply:
ok, thanks a lot for clarifying! it can't have been easy to implement all this :)